File: backend/app/services/room_service.py
"""Room management service"""

from uuid import UUID, uuid4
from app.core.database import Database
from app.models.schemas import Room, User


class RoomService:
    def __init__(self, db: Database):
        self.db = db

    async def create(self, name: str, host_id: UUID, **kwargs) -> Room | None:
        room_id = uuid4()
        row = await self.db.fetchrow(
            """
            INSERT INTO rooms (
                id,
                name,
                description,
                media_url,
                media_title,
                episode,
                anilist_id,
                mal_id,
                is_private,
                max_users,
                host_id
            )
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
            RETURNING *
            """,
            room_id,
            name,
            kwargs.get("description", ""),
            kwargs.get("media_url"),
            kwargs.get("media_title"),
            kwargs.get("episode"),
            kwargs.get("anilist_id"),
            kwargs.get("mal_id"),
            kwargs.get("is_private", False),
            kwargs.get("max_users", 10),
            host_id,
        )
        if row:
            await self.db.execute(
                "INSERT INTO room_members (room_id, user_id, role) VALUES ($1, $2, 'host')",
                room_id, host_id,
            )
            return await self.get(room_id)
        return None

    async def get(self, room_id: UUID) -> Room | None:
        row = await self.db.fetchrow(
            """
            SELECT
                r.*,
                p.id AS host_user_id,
                p.username AS host_username,
                p.display_name AS host_display_name,
                p.avatar_url AS host_avatar_url,
                p.status AS host_status,
                p.created_at AS host_created_at,
                (
                    SELECT COUNT(*)
                    FROM room_members rm
                    WHERE rm.room_id = r.id AND rm.is_banned = false
                ) AS participant_count
            FROM rooms r
            JOIN profiles p ON p.id = r.host_id
            WHERE r.id = $1
            """,
            room_id,
        )
        if not row:
            return None
        members = await self.get_members(room_id)
        host = User(
            id=row["host_user_id"],
            username=row["host_username"],
            display_name=row["host_display_name"],
            avatar_url=row["host_avatar_url"],
            status=row["host_status"],
            created_at=row["host_created_at"],
        )
        room_data = dict(row)
        for k in [
            "host_user_id",
            "host_username",
            "host_display_name",
            "host_avatar_url",
            "host_status",
            "host_created_at",
        ]:
            room_data.pop(k, None)
        return Room(**room_data, host=host, participant_count=row["participant_count"], members=members)

    async def get_members(self, room_id: UUID) -> list[User]:
        rows = await self.db.fetch(
            """
            SELECT p.id, p.username, p.display_name, p.avatar_url, p.status
            FROM room_members rm
            JOIN profiles p ON p.id = rm.user_id
            WHERE rm.room_id = $1 AND rm.is_banned = false
            """,
            room_id,
        )
        return [User(**dict(r)) for r in rows]

    async def join(self, room_id: UUID, user_id: UUID) -> bool:
        exists = await self.db.fetchrow(
            "SELECT 1 FROM room_members WHERE room_id = $1 AND user_id = $2",
            room_id, user_id,
        )
        if exists:
            return True
        row = await self.db.fetchrow("SELECT max_users FROM rooms WHERE id = $1", room_id)
        if not row:
            return False
        count = await self.db.fetchval(
            "SELECT COUNT(*) FROM room_members WHERE room_id = $1",
            room_id,
        )
        if count >= row["max_users"]:
            return False
        await self.db.execute(
            "INSERT INTO room_members (room_id, user_id, role) VALUES ($1, $2, 'member')",
            room_id, user_id,
        )
        return True

    async def leave(self, room_id: UUID, user_id: UUID):
        await self.db.execute(
            "DELETE FROM room_members WHERE room_id = $1 AND user_id = $2",
            room_id, user_id,
        )

    async def update_room(self, room_id: UUID, **fields) -> Room | None:
        if not fields:
            return await self.get(room_id)
        set_clause = ", ".join(f"{k} = ${i + 1}" for i, k in enumerate(fields))
        values = list(fields.values())
        values.append(room_id)
        row = await self.db.fetchrow(
            f"UPDATE rooms SET {set_clause}, updated_at = NOW() WHERE id = ${len(values)} RETURNING *",
            *values,
        )
        if not row:
            return None
        return await self.get(room_id)

    async def get_public(self, limit: int = 50, offset: int = 0) -> list[Room]:
        rows = await self.db.fetch(
            """
            SELECT
                r.*,
                p.id AS host_user_id,
                p.username AS host_username,
                p.display_name AS host_display_name,
                p.avatar_url AS host_avatar_url,
                p.status AS host_status,
                p.created_at AS host_created_at,
                (
                    SELECT COUNT(*)
                    FROM room_members rm
                    WHERE rm.room_id = r.id AND rm.is_banned = false
                ) AS participant_count
            FROM rooms r
            JOIN profiles p ON p.id = r.host_id
            WHERE r.is_private = false
            ORDER BY r.created_at DESC
            LIMIT $1 OFFSET $2
            """,
            limit,
            offset,
        )
        items: list[Room] = []
        for r in rows:
            host = User(
                id=r["host_user_id"],
                username=r["host_username"],
                display_name=r["host_display_name"],
                avatar_url=r["host_avatar_url"],
                status=r["host_status"],
                created_at=r["host_created_at"],
            )
            room_data = dict(r)
            for k in [
                "host_user_id",
                "host_username",
                "host_display_name",
                "host_avatar_url",
                "host_status",
                "host_created_at",
            ]:
                room_data.pop(k, None)
            items.append(Room(**room_data, host=host, participant_count=r["participant_count"], members=[]))
        return items

    async def count_public(self) -> int:
        return await self.db.fetchval("SELECT COUNT(*) FROM rooms WHERE is_private = false")

File: apps/api/src/services/room.service.ts
import { supabase } from '../lib/supabase';
import { redisService } from './redis.service';
import { Room, RoomMember } from '@syncsaga/shared';
import { logger } from '../lib/logger';
import * as bcrypt from 'bcrypt';

const BCRYPT_ROUNDS = 12;
const MAX_ROOM_USERS = 50;

export class RoomService {
  /**
   * Create a new room with hashed password if private.
   */
  async createRoom(data: {
    name: string;
    description?: string;
    isPrivate?: boolean;
    maxUsers?: number;
    hostId: string;
    animeTitle?: string;
    animeMediaId?: number;
    password?: string;
  }): Promise<Room | null> {
    const insertData: Record<string, unknown> = {
      name: data.name,
      description: data.description,
      is_private: data.isPrivate ?? false,
      max_users: Math.min(data.maxUsers ?? 10, MAX_ROOM_USERS),
      host_id: data.hostId,
    };

    // Hash password if provided
    if (data.password) {
      insertData.password = await bcrypt.hash(data.password, BCRYPT_ROUNDS);
    }

    if (data.animeTitle) insertData.anime_title = data.animeTitle;
    if (data.animeMediaId) insertData.anime_media_id = data.animeMediaId;

    const { data: room, error } = await supabase.from('rooms').insert(insertData).select().single();
    if (error) {
      logger.error({ err: error }, 'Failed to create room');
      return null;
    }

    // Add host as first member
    const { error: memberError } = await supabase
      .from('room_members')
      .insert({ room_id: room.id, user_id: data.hostId, role: 'host' });

    if (memberError) {
      logger.error({ err: memberError, roomId: room.id }, 'Failed to add host as room member');
    }

    return room as Room;
  }

  /**
   * Get a room by ID with members.
   * Password is excluded from the returned object.
   */
  async getRoom(roomId: string): Promise<(Room & { members: RoomMember[] }) | null> {
    const { data: room, error } = await supabase
      .from('rooms')
      .select('*')
      .eq('id', roomId)
      .single();

    if (error || !room) return null;

    const { data: members } = await supabase
      .from('room_members')
      .select('*')
      .eq('room_id', roomId)
      .eq('is_banned', false);

    // Never return the password hash to the client
    const { password: _pwd, ...roomWithoutPassword } = room;
    return { ...roomWithoutPassword, members: (members || []) as RoomMember[] } as Room & { members: RoomMember[] };
  }

  /**
   * Join a room with password verification and capacity check.
   * Uses a distributed lock to prevent race conditions on max_users.
   */
  async joinRoom(roomId: string, userId: string, password?: string): Promise<boolean> {
    const room = await this.getRoom(roomId);
    if (!room) return false;

    // Check if already a member
    const isMember = room.members.some(m => m.user_id === userId);
    if (isMember) return true;

    // Check capacity
    if (room.members.length >= room.max_users) return false;

    // Verify password for private rooms
    if (room.is_private) {
      // Fetch the password hash separately (not returned by getRoom)
      const { data: roomData } = await supabase
        .from('rooms')
        .select('password')
        .eq('id', roomId)
        .single();

      if (!roomData?.password) return false;
      const passwordValid = await bcrypt.compare(password || '', roomData.password);
      if (!passwordValid) return false;
    }

    // Use distributed lock to prevent concurrent join race
    const lockId = await redisService.acquireLock(`room:join:${roomId}`, 2000);
    if (!lockId) {
      // Another join in progress — retry once
      await new Promise(resolve => setTimeout(resolve, 100));
      const retryLock = await redisService.acquireLock(`room:join:${roomId}`, 2000);
      if (!retryLock) return false;
      try {
        // Re-check capacity under lock
        const { count } = await supabase
          .from('room_members')
          .select('*', { count: 'exact', head: true })
          .eq('room_id', roomId)
          .eq('is_banned', false);

        if ((count || 0) >= room.max_users) return false;

        const { error } = await supabase
          .from('room_members')
          .insert({ room_id: roomId, user_id: userId, role: 'member' });
        return !error;
      } finally {
        await redisService.releaseLock(`room:join:${roomId}`, retryLock);
      }
    }

    try {
      const { error } = await supabase
        .from('room_members')
        .insert({ room_id: roomId, user_id: userId, role: 'member' });
      return !error;
    } finally {
      await redisService.releaseLock(`room:join:${roomId}`, lockId);
    }
  }

  /**
   * Leave a room and clean up Redis presence.
   */
  async leaveRoom(roomId: string, userId: string): Promise<void> {
    await supabase
      .from('room_members')
      .delete()
      .eq('room_id', roomId)
      .eq('user_id', userId);
    await redisService.removeUserFromRoom(roomId, userId);
  }

  /**
   * Update room state in both database and Redis cache.
   */
  async updateRoomState(roomId: string, state: Partial<Room>): Promise<void> {
    const allowed = ['playback_state', 'current_timestamp', 'playback_speed', 'current_episode', 'current_episode_number', 'updated_at'] as const;
    const safeState: Record<string, unknown> = {};
    for (const key of allowed) {
      if (key in state) safeState[key] = (state as Record<string, unknown>)[key];
    }
    const { error } = await supabase.from('rooms').update(safeState).eq('id', roomId);
    if (error) {
      logger.error({ err: error, roomId }, 'Failed to update room in database');
    }
    await redisService.setRoomState(roomId, safeState);
  }

  /**
   * Get public rooms with pagination.
   */
  async getPublicRooms(limit = 20, offset = 0): Promise<Room[]> {
    const { data } = await supabase
      .from('rooms')
      .select('*')
      .eq('is_private', false)
      .order('created_at', { ascending: false })
      .range(offset, offset + limit - 1);
    return (data || []) as Room[];
  }
}

export const roomService = new RoomService();
